File: This_is_conding_test/practice/topology.py
# 7 8
# 1 2
# 1 5
# 2 3
# 2 6
# 3 4
# 4 7
# 5 6
# 6 4

# 1 2 5 3 6 4 7

from collections import deque

N, M = map(int, input().split())

indegree = [0] * (N+1)
graph = [[] for _ in range(N+1)]

for _ in range(M):
    startNode, endNode = map(int,input().split())
    graph[startNode].append(endNode)
    indegree[endNode] += 1

def topologySort():
    q = deque()
    result = []

    for i in range(1, N + 1):
        if indegree[i] == 0:
            q.append(i)

    while q:
        node = q.popleft()
        result.append(node)

        for i in graph[node]:
            indegree[i] -= 1
            if indegree[i] == 0:
                q.append(i)

    print(result, end = " ")

topologySort()

File: JS/topology.js
const fs = require("fs");
const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `7 8
1 2
1 5
2 3
2 6
3 4
4 7
5 6
6 4
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

let [N, M] = input().split(" ").map(Number);

let graph = Array(N + 1)
  .fill(null)
  .map(() => Array());

let indegree = Array(N + 1).fill(0);

for (let i = 0; i < M; i++) {
  const [startNode, endNode] = input().split(" ").map(Number);
  graph[startNode].push(endNode);
  indegree[endNode]++;
}
function topologySort() {
  q = [];
  result = [];

  for (let i = 1; i < N + 1; i++) {
    if (indegree[i] === 0) q.push(i);
  }

  while (q.length > 0) {
    nowNode = q.shift();
    result.push(nowNode);

    for (let i of graph[nowNode]) {
      indegree[i]--;
      if (indegree[i] === 0) {
        q.push(i);
      }
    }
  }

  console.log(result);
}

topologySort();
